### PR TITLE
tests: posix: common: separate xsi_threads_ext into standalone test

### LINKED ISSUE #81741: tests: posix: common: split xsi_threads_ext into separate test
**Is your enhancement proposal related to a problem? Please describe.**
Make the xsi_threads_ext-related tests in `tests/posix/common` into a standalone test.

**Describe the solution you'd like**
The obvious way to group xsi_threads_ext-related testsuites is by `XSI_THREADS_EXT` [Option Group](https://pubs.opengroup.org/onlinepubs/9799919799/xrat/V4_subprofiles.html).

**Describe alternatives you've considered**
The way the testsuite has been designed historically.

It's probably fine to group the `XSI_THREADS_EXT tests into this suite as well.

**Additional context**
See #79454 for more info

### PR DESCRIPTION
**Pull request by Pancakem**
Fixes #81741

posix.common contains testsuites that can be separated into smaller groups of tests. This change moves tests in the xsi_threads_ext option group into a singular testsuite tests/posix/xsi_threads_ext app directory.

### CODE REVIEW

File: tests/posix/xsi_threads_ext/testcase.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    - posix
+    - xsi_threads_ext
+  # 1 tier0 platform per supported architecture
+  platform_key:
+    - arch
+    - simulation
+tests:
+  portability.posix.xsi_threads_ext:
+    min_flash: 64
+    min_ram: 32
```
Comment (cfriedt) [inline tests/posix/xsi_threads_ext/testcase.yaml:13]:
```suggestion
  platform_key:
    - arch
    - simulation
  min_flash: 64
  min_ram: 32
tests:
  portability.posix.xsi_threads_ext: {}
```

**Review by cfriedt — CHANGES_REQUESTED** [review]

File: tests/posix/common/src/pthread_attr.c [hunk 2]
```diff
@@ -287,97 +287,6 @@ ZTEST(pthread_attr, test_pthread_attr_setschedpolicy)
...
-		zassert_ok(pthread_attr_setstack(&attr, (void *)stack, 2 * stacksize));
-		/* ensure we read back the same values as we specified */
-		zassert_ok(pthread_attr_getstack(&attr, &new_stackaddr, &new_stacksize));
-		zassert_equal(new_stackaddr, (void *)stack);
-		zassert_equal(new_stacksize, 2 * stacksize);
-		can_create_thread(&attr);
-	}
-}
-
 ZTEST(pthread_attr, test_pthread_attr_getstacksize)
```
Comment (cfriedt) [inline tests/posix/common/src/pthread_attr.c:290]: I think that `test_pthread_attr_getstacksize` and `test_pthread_attr_setstacksize` probably also need to be moved to the `xsi_threads_ext` group

**Review by cfriedt — CHANGES_REQUESTED** [review]
Please also relocate `test_pthread_attr_getstacksize` and `test_pthread_attr_setstacksize` in this PR. They aren't in [XSI_THREADS_EXT](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html) today, but they were also deprecated a while ago, IIRC.

File: tests/posix/xsi_threads_ext/src/main.c [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+#define BIOS_FOOD 0xB105F00D
+
+static bool attr_valid;
+static pthread_attr_t attr;
+static const pthread_attr_t uninit_attr;
+static bool detached_thread_has_finished;
+
+/* TODO: this should be optional */
+#define STATIC_THREAD_STACK_SIZE (MAX(1024, PTHREAD_STACK_MIN + CONFIG_TEST_EXTRA_STACK_SIZE))
```
Comment (cfriedt) [inline tests/posix/xsi_threads_ext/src/main.c:21]: Where is `CONFIG_TEST_EXTRA_STACK_SIZE` set in the `common` test suite?

File: tests/posix/xsi_threads_ext/src/main.c [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+		 */
+		zassert_ok(pthread_attr_setstack(&attr, stackaddr, stacksize - 1));
+		/* ensure we read back the same values as we specified */
+		zassert_ok(pthread_attr_getstack(&attr, &new_stackaddr, &new_stacksize));
+		zassert_equal(new_stackaddr, stackaddr);
+		zassert_equal(new_stacksize, stacksize - 1);
+		can_create_thread(&attr);
+	}
+
+	if (IS_ENABLED(DYNAMIC_THREAD_ALLOC)) {
```
Comment (cfriedt) [inline tests/posix/xsi_threads_ext/src/main.c:155]:
```suggestion
	if (IS_ENABLED(CONFIG_DYNAMIC_THREAD_ALLOC)) {
```

Where is `CONFIG_DYNAMIC_THREAD_ALLOC` set in the `common` test suite and why doesn't it appear in this test suite?

File: tests/posix/xsi_threads_ext/src/main.c [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+	/* qemu_x86 seems to be unable to set thread stacks to be anything less than 4096 */
+	if (!IS_ENABLED(CONFIG_X86)) {
+		zassert_ok(pthread_attr_setstacksize(&attr, stacksize - 1));
+		/* ensure we can read back the values we just set */
+		zassert_ok(pthread_attr_getstacksize(&attr, &new_stacksize));
+		zassert_equal(new_stacksize, stacksize - 1);
+		can_create_thread(&attr);
+	}
+
+	if (IS_ENABLED(CONFIG_DYNAMIC_THREAD_ALLOC)) {
```
Comment (cfriedt) [inline tests/posix/xsi_threads_ext/src/main.c:253]: Where is `CONFIG_DYNAMIC_THREAD_ALLOC` set in the `common` test suite and why doesn't it appear in this test suite?

**Review by cfriedt — CHANGES_REQUESTED** [review]
I think we will need an additional testcase configuration in the new test to cover `CONFIG_DYNAMIC_THREAD_ALLOC` like in the `common` test suite.

Also, not sure if the config for extra stack size was specified in a board file or Kconfig, but that should be checked.

File: tests/posix/xsi_threads_ext/src/main.c (hunk 3 above)
Comment (Pancakem) [reply tests/posix/xsi_threads_ext/src/main.c:21]: It has been set in both https://github.com/zephyrproject-rtos/zephyr/blob/5dfd41e51a86c75d156241b7566ad4c1c1dce2d5/tests/posix/common/testcase.yaml#L71 and https://github.com/zephyrproject-rtos/zephyr/blob/5dfd41e51a86c75d156241b7566ad4c1c1dce2d5/tests/posix/common/prj.conf#L13. 
I will set it.

File: tests/posix/xsi_threads_ext/testcase.yaml [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    extra_configs:
+      - CONFIG_NEWLIB_LIBC=y
+  portability.posix.xsi_threads_ext.picolibc:
+    tags: picolibc
+    filter: CONFIG_PICOLIBC_SUPPORTED
+    extra_configs:
+      - CONFIG_PICOLIBC=y
+  portability.posix.xsi_threads_ext.dynamic_stack:
+    extra_configs:
+      - CONFIG_TEST_EXTRA_STACK_SIZE=16384
```
Comment (Pancakem) [inline tests/posix/xsi_threads_ext/testcase.yaml:28]: I am sure whether this is necessary.

File: tests/posix/xsi_threads_ext/src/main.c (hunk 3 above)
Comment (cfriedt) [reply tests/posix/xsi_threads_ext/src/main.c:21]: It was added here, and likely completely unnecessary for this test

https://github.com/zephyrproject-rtos/zephyr/commit/43f917111b30790801824cab050a758817fd03f2

File: tests/posix/xsi_threads_ext/src/main.c [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+	}
+}
+
+static void before(void *arg)
+{
+	ARG_UNUSED(arg);
+
+	zassert_ok(pthread_attr_init(&attr));
+	/* TODO: pthread_attr_init() should be sufficient to initialize a thread by itself */
+	zassert_ok(pthread_attr_setstack(&attr, &static_thread_stack, STATIC_THREAD_STACK_SIZE));
```
Comment (cfriedt) [inline tests/posix/xsi_threads_ext/src/main.c:268]:
```suggestion
if (!IS_ENABLED(CONFIG_DYNAMIC_THREAD_ALLOC)) {
	zassert_ok(pthread_attr_setstack(&attr, &static_thread_stack, STATIC_THREAD_STACK_SIZE));
}
```

File: tests/posix/xsi_threads_ext/testcase.yaml (hunk 6 above)
Comment (cfriedt) [inline tests/posix/xsi_threads_ext/testcase.yaml:28]:
```suggestion
  portability.posix.xsi_threads_ext.static_stack:
    extra_configs:
      - CONFIG_DYNAMIC_THREAD_ALLOC=n
```

**Review by cfriedt — CHANGES_REQUESTED** [review]

File: tests/posix/xsi_threads_ext/testcase.yaml (hunk 6 above)
Comment (cfriedt) [reply tests/posix/xsi_threads_ext/testcase.yaml:28]: It isn't

File: tests/posix/xsi_threads_ext/src/main.c [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+#define BIOS_FOOD 0xB105F00D
+
+static bool attr_valid;
+static pthread_attr_t attr;
+static const pthread_attr_t uninit_attr;
+static bool detached_thread_has_finished;
+
+/* TODO: this should be optional */
+#define STATIC_THREAD_STACK_SIZE (MAX(1024, PTHREAD_STACK_MIN + CONFIG_TEST_EXTRA_STACK_SIZE))
+static K_THREAD_STACK_DEFINE(static_thread_stack, STATIC_THREAD_STACK_SIZE);
```
Comment (cfriedt) [inline tests/posix/xsi_threads_ext/src/main.c:22]:
```suggestion
/* This should be discarded by the linker, in this specific testsuite, if CONFIG_DYNAMIC_THREAD_ALLOC is not set */
#define STATIC_THREAD_STACK_SIZE (MAX(1024, PTHREAD_STACK_MIN + CONFIG_TEST_EXTRA_STACK_SIZE))
static K_THREAD_STACK_DEFINE(static_thread_stack, STATIC_THREAD_STACK_SIZE);
```

File: tests/posix/xsi_threads_ext/src/main.c (hunk 5 above)
Comment (cfriedt) [reply tests/posix/xsi_threads_ext/src/main.c:253]: It should be in prj.conf now

**Review by cfriedt — CHANGES_REQUESTED** [review]

**Review by cfriedt — APPROVED** [review]

**Review by kartben — APPROVED** [review]

**Pull request merged**